In [65]:
import itertools
import pandas as pd
import numpy as np

def dice_probability_A(sides, n_rolls, target):
    # List out all possible rolls
    rolls = list(itertools.product(range(1, sides+1), repeat=n_rolls))

    # Find the sum of all possible rolls
    sums = np.array([sum(roll) for roll in rolls])

    # Find the proportion of those sums that are greater than or equal to the target
    return np.mean(sums >= target)


def dice_probability_B(sides, n_rolls, target):
    # List out all possible rolls
    rolls = list(itertools.product(range(1, sides+1), repeat=n_rolls))

    # Get the total number of rolls that contain a number greater than or equal to the target
    # Dynamic for n_rolls
    target_rolls = [[str(a) for a in roll if target <= a] for roll in rolls]
    
    # Gets rid of excess spaces and only has numbers greater than or equal to the target
    target_rolls = ' '.join([''.join(roll) for roll in target_rolls if roll != '' ]).split()

    # Divides the number of rolls greater than or equal to the target against the total rolls
    return len(target_rolls) / len(rolls)

# Situation A
Calculating the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x. 

This calculation should only be implemented for m = 1 through 4 and n is at most 20

In [66]:
d_20 = dice_probability_A(20, 1, 17) * 100
print(f'Roll 1d20 and get a result of 17 or more: {d_20:.2f}%')

d_4 = dice_probability_A(4, 3, 10) * 100
print(f'Roll 3d4 and get a result of 10 or more: {d_4:.2f}%')

Roll 1d20 and get a result of 17 or more: 20.00%
Roll 3d4 and get a result of 10 or more: 15.62%


# Situation B
Determining the probability of rolling m dice, each with n sides, and having at least one die show a value greater than or equal to x

In [72]:
d_6 = dice_probability_B(6, 4, 6) * 100
print(f'Roll 4d6 and have at least one dice show a 6: {d_6:.2f}%')

Roll 4d6 and have at least one dice show a 6: 51.77%


The highest probability will be rolling 4d6 and getting at least one 6 with a 51.77% chance of happening compared to a 20% chance of rolling 1d20 and getting a 17 or higher and 15.62% change of rolling 3d4 and getting 10 or more